<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>TRENA: computational prediction of gene regulation • trena</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="TRENA: computational prediction of gene regulation">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">trena</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.7.14</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/TReNA_Vignette.html">A Brief Introduction to TReNA</a>
    </li>
    <li>
      <a href="../articles/caseStudyFour.html">Case Study Four: a novel regulator of GATA2 in erythropoieis?</a>
    </li>
    <li>
      <a href="../articles/caseStudyOne.html">Case Study One: attempt to reproduce known regulation of NFE2 by GATA1 in bulk RNA-seq</a>
    </li>
    <li>
      <a href="../articles/caseStudyThree.html">Case Study Three attempt to reproduce known regulation of NFE2 by GATA1 in bulk RNA-seq</a>
    </li>
    <li>
      <a href="../articles/caseStudyTwo.html">Case Study Two reproduces known regulation of NFE2 by GATA1 in erytrhopoiesis RNA-seq</a>
    </li>
    <li>
      <a href="../articles/overview.html">TRENA: computational prediction of gene regulation</a>
    </li>
    <li>
      <a href="../articles/simple.html">Knitr/R Markdown: explore output controls</a>
    </li>
    <li>
      <a href="../articles/tiny.html">An example Knitr/R Markdown document</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>TRENA: computational prediction of gene regulation</h1>
            
      
      
      <div class="hidden name"><code>overview.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>The regulation of target genes by their transcription factors is complex and incompletely understood. Multiple signals involving core promoters, distal enhancers, epigenetic controls on chromatin accessibility, stochastic and cooperative binding on different times scales are all involved.</p>
<p>Many hundreds of human TF/target gene relationships <em>are</em> known, however. These relationships are primarily from well-studied systems and from research into human diseases. Hundreds of thousands of TF/target gene relationships exist in human cells. There is an urgent need for laboratory and computational methods which predict and validate the many that remain unknown.</p>
<p><em>trena</em> is one such computational method. It’s predictive success depends upon the judicious combination of two predictors, each of which in isolation has little predictive value:</p>
<ol style="list-style-type: decimal">
<li>correlated gene expression of TF and target gene</li>
<li>actual or predicted DNA binding of the TF in regulatory DNA regions associated with the target gene</li>
</ol>
<p>We will show that these predictors, when applied in tandem to high quality gene expression, genomic and epigenomic data, can 1) recapitulate known TF/target gene relationships, and 2) predict new ones sufficiently plausible to warrant small-scale laboratory valdidation, and 3) specify useful genome-scale tissue-specific regulatory networks.</p>
</div>
<div id="methodological-overview" class="section level1">
<h1 class="hasAnchor">
<a href="#methodological-overview" class="anchor"></a>Methodological Overview</h1>
<p>There are two stages in every trena analysis:</p>
<ol style="list-style-type: decimal">
<li>Select a relevant set of transcription factors</li>
<li>Score the gene expression association of each TF with the target gene using an ensemble of the independent feature selectors trena provides: lasso, ridge, xgboost, randomForest pearson &amp; spearman correlation.</li>
</ol>
<div id="step-one-select-candidate-transcription-factors" class="section level3">
<h3 class="hasAnchor">
<a href="#step-one-select-candidate-transcription-factors" class="anchor"></a>Step One: select candidate transcription factors</h3>
<p>Step one can be accomplished in many ways. A naive approach uses all 1663 human genes currently annotated by the GeneOntology project to the molecular function, “DNA-binding transcription factor activity”. A more conservative choice is to use the 558 human genes for which Jaspar2018 and HOCOMOCO provide binding motifs. We will demonstrate these two annotation-based strategies below and then proceed to use more discriminating approaches:</p>
<ol style="list-style-type: decimal">
<li>Assume a traditional cis-regulatory promoter region for the target gene (say, +/- 5kb from the TSS).</li>
<li>Obtain cis-regulatory promoters and enhancers from public databaes or - best of all - by cell type specific experimental data for the system under study</li>
<li>Select transcription factors with motifs which match DNA sequence in these regulatory regions. Each match can be quantified in (at least) two ways: by motif/sequence match, and by the evolutionary conservation of that matching sequence.</li>
</ol>
<p>Each of these strategies has both strengths weaknesses. The naive 1663 GO annotation gene strategy will identify TFs (or co-factors) which may be active and functional but for which a binding motif is not neccesarily known. The 558 motif TF strategy similarly makes no assumptions about regulatory regions, but probably is enriched for better studied TFs. In the absence of good data about the target genes’s actual regulatory regions, and binding, either of these inclusive approaches can be useful.</p>
<p>However, there is an emerging consensus (Cusanovich 2014 &amp; etc) that a great deal of actually bound TFs, as detected by ChIP-seq, in known regulatory regions, are not functional. Their knockdown or silencing has no effect on the expression of the target gene. Thus the primary challenge for the first stage of trena - TF selection - is how to avoid considering too many extraneous TF candidates.</p>
<p>We adopt these strategies for TF selection:</p>
<ol style="list-style-type: decimal">
<li>select only TFs with motifs with very high DNA matching scores</li>
<li>occuring in non-coding DNA sequence that is highly conserved over 100 million years</li>
<li>when avaiable, as with the increasingly available single-cell ATAC-seq, further filter the TFs by requring that condition-specific chromatin is open</li>
</ol>
<p>The stringency of strategies 1 and 2 will, in any given study, exclude functional and relevant binding sites, and their cognate TFs from consideration in the trena regulatory model. This loss is more than offset, as we will show, by the credibility of the models which emerge from that stringency, based on the assumption that (again, following Cusanovich 2014) that since each human gene has a median TF count of 35, that high fidelity motif match in highly conserved regions of open chromatin identifies at least some TFs of interest - and, we speculate, that these will often include the oldest and most central TFs for each target gene.</p>
</div>
<div id="step-two-run-the-ensemble-of-feature-selectors" class="section level3">
<h3 class="hasAnchor">
<a href="#step-two-run-the-ensemble-of-feature-selectors" class="anchor"></a>Step Two: Run the ensemble of feature selectors</h3>
<p>After appropriate normalization, the expression of the TFs selected in step one are used to predict the expression of the target gene, using a variety of algorithms:</p>
<ul>
<li>lasso</li>
<li>lassoPValue</li>
<li>ridge regression</li>
<li>pearson correlation</li>
<li>spearman correlation</li>
<li>random forest</li>
<li>xgboost</li>
</ul>
<p>The resulting “trena model” is an R data.frame containing the score using each algorithm of each TF as a predictor of the gene expression of the target gene. We as yet have no single settled heuristic method to aggregate these scores. Individucal analysts evolve their own informal aggregation scheme. For example: giving weight and preference to the random forest score while also heeding spearman correlation. Quite commonly, as will be seen in the case studies, the various methods produce compatible predictions.</p>
<p>The first criterion of value for a computational tool such as trena is its ability to reliably reproduce existing knowledge. Here we accomplish that with four case studies, first by demonstrating failure with bulk GTEx blood RNA-seq, then success with three applications of stage-specific erythropoiesis RNA-seq and ATAC-seq data.</p>
</div>
<div id="case-studies" class="section level3">
<h3 class="hasAnchor">
<a href="#case-studies" class="anchor"></a>Case Studies</h3>
<ul>
<li>
<a href="caseStudyOne.html">Case Study 1</a> An informative failure: trena with GTEx bulk blood expression</li>
<li>
<a href="caseStudyTwo.html">Case Study 2</a> Successful recapitulation: trena + Corces et al, 2016, RNA-seq on FACS-separated erythropoiesis stages, GSE74246</li>
<li>
<a href="caseStudyThree.html">Case Study 3</a> Successful recapituation: more recent Brand lab (similar to Corces et al), GSE118537</li>
<li>
<a href="caseStudyFour.html">Case Study 4</a> Discover a novel regulator of GATA2?: trena + Brand lab + scATAC-seq</li>
</ul>
<!--
This image summarizes the NFE2 genome neighborhood and binding sites for putative regulators identified
by trena:

  1. phast7: conserved DNA sequence, in red, on a scale of 0-1, across 7 species: opossum, dog, rat, mouse, rhesus
     monkey, chimpanzee and human.  Highly conserved sequence in non-coding regions is likely to be functional,
     that is, to play a role in regulating a nearby gene, and possibly a TF binding site.

  2. atac combined: the union of 12 scATAC-seq datasets, collected in an erythropoiesis time course
     Gillespie et al, https://www.biorxiv.org/content/10.1101/812123v1

  3. TAL1, KLF1, GATA1,  etc:  high-scoring motif matches for top regulating TFs predicted by trena
     using RNA-seq expression data from Gillespie et al.

Here is the high-res model as scored by trena, followed by the genome view.



          betaLasso   lassoPValue   pearsonCoeff   rfScore   betaRidge   spearmanCoeff   xgboost   bindingSites
-------  ----------  ------------  -------------  --------  ----------  --------------  --------  -------------
TAL1           0.34      0.00e+00           0.92      1.83        0.08            0.93      0.09              1
KLF1           0.08      1.00e-07           0.85      2.16        0.04            0.96      0.79              1
GATA1          0.11      0.00e+00           0.84      1.09        0.09            0.90      0.00              1
SP2            0.00      8.53e-01           0.80      0.54        0.13            0.81      0.00              3
KLF3           0.08      2.50e-06           0.78      0.83        0.08            0.77      0.04              1
SP1            0.00      4.73e-02           0.69      0.38        0.16            0.78      0.04              2
FLI1           0.00      4.77e-01          -0.66      0.13       -0.01           -0.62      0.00              1
KLF8           0.00      7.33e-01          -0.65      0.22       -0.02           -0.64      0.00              1
RARA           0.00      9.01e-01          -0.64      0.18       -0.02           -0.51      0.00              1
SPI1           0.00      9.73e-01          -0.58      0.07       -0.01           -0.52      0.00              1
POU3F1         0.00      8.78e-01          -0.54      0.04       -0.02           -0.50      0.00              2
CEBPA          0.00      7.42e-01          -0.53      0.23       -0.01           -0.62      0.00              1


<img src="/Users/paul/github/trena/vignettes/igvR-nfe2-regulation-noGeneHancer.png" width="852" />


## Correlated gene expression of all known TFs to the target gene

We do this with three (four) sources of gene expression data

 1.  GTEx Whole Blood

### GTEx Blood


### Corces et al; Lineage-specific and single-cell chromatin accessibility charts human hematopoiesis and leukemia evolution.
in 13 human proimary blood cell types that span the hamatopoietic hierarchy"

see al
    https://www.ncbi.nlm.nih.gov/pubmed/27526324

see also: Integrated Single-Cell Analysis Maps the Continuous Regulatory Landscape of Human Hematopoietic Differentiation
Jason D.Buenrostro et al

### Brand erythropoiesis time-course

## Further restrict TFs to those which fall in highly conserved (phast7 >= 0.5) regions

##  as above, but with RNA collected from multiple stages of erythropoiesis
  5. use RNA-seq and scATAC-seq from Gillespie et al.  The ATAC-seq track is the union of
    for 20 samples across 12 erythropoiesis development timepoints.


# Journal Articles Reporting Regulators of NFE2  

  - Bose, Francesca, et al. "Functional interaction of CP2 with GATA-1 in the regulation of erythroid promoters."
    Molecular and cellular biology 26.10 (2006): 3942-3954.

  - Ding, Ya‐li, et al. "Over‐expression of EDAG in the myeloid cell line 32D: Induction of GATA‐1 expression
    and erythroid/megakaryocytic phenotype." Journal of cellular biochemistry 110.4 (2010): 866-874.

  - many more to be added...




# Session Info


```r
sessionInfo()
```

```
## R version 3.6.1 (2019-07-05)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS Mojave 10.14.5
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.3       png_0.1-7        rprojroot_1.3-2  crayon_1.3.4    
##  [5] digest_0.6.22    assertthat_0.2.1 MASS_7.3-51.4    R6_2.4.1        
##  [9] backports_1.1.5  magrittr_1.5     evaluate_0.14    highr_0.8       
## [13] stringi_1.4.3    rlang_0.4.1      fs_1.3.1         rmarkdown_1.17  
## [17] pkgdown_1.4.1    desc_1.2.0       tools_3.6.1      stringr_1.4.0   
## [21] yaml_2.2.0       xfun_0.11        compiler_3.6.1   memoise_1.1.0   
## [25] htmltools_0.4.0  knitr_1.26
```


![knitr logo](http://yihui.name/knitr/images/knit-logo.png)
-->
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#methodological-overview">Methodological Overview</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Paul Shannon.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
