<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Inferring Transcriptional Regulation with TReNA — trena-package • trena</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Inferring Transcriptional Regulation with TReNA — trena-package" />
<meta property="og:description" content="`trena` provides a framework for using gene expression data to infer relationships
between a target gene and a set of transcription factors. It does so using a
several classes and their associated methods, briefly documented below" />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">trena</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.7.14</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/TReNA_Vignette.html">A Brief Introduction to TReNA</a>
    </li>
    <li>
      <a href="../articles/caseStudyFour.html">Case Study Four: a novel regulator of GATA2 in erythropoieis?</a>
    </li>
    <li>
      <a href="../articles/caseStudyOne.html">Case Study One: attempt to reproduce known regulation of NFE2 by GATA1 in bulk RNA-seq</a>
    </li>
    <li>
      <a href="../articles/caseStudyThree.html">Case Study Three attempt to reproduce known regulation of NFE2 by GATA1 in bulk RNA-seq</a>
    </li>
    <li>
      <a href="../articles/caseStudyTwo.html">Case Study Two reproduces known regulation of NFE2 by GATA1 in erytrhopoiesis RNA-seq</a>
    </li>
    <li>
      <a href="../articles/overview.html">TRENA: computational prediction of gene regulation</a>
    </li>
    <li>
      <a href="../articles/simple.html">Knitr/R Markdown: explore output controls</a>
    </li>
    <li>
      <a href="../articles/tiny.html">An example Knitr/R Markdown document</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Inferring Transcriptional Regulation with TReNA</h1>
    
    <div class="hidden name"><code>trena-package.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>`trena` provides a framework for using gene expression data to infer relationships
between a target gene and a set of transcription factors. It does so using a
several classes and their associated methods, briefly documented below</p>
    </div>



    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>#'
Solver Class Objects</p>
<p>The <code><a href='Solver-class.html'>Solver</a></code> class is a base class used within `trena`. A particular
<code><a href='Solver-class.html'>Solver</a></code> object also contains the name of the selected solver and
dispatches the correct feature selection method when <code>run</code> is called on the
object. It is inherited by all the following subclasses, representing the
different feature selection methods: <code><a href='BayesSpikeSolver.html'>BayesSpikeSolver</a></code>,
<code><a href='EnsembleSolver.html'>EnsembleSolver</a></code>, <code><a href='LassoPVSolver.html'>LassoPVSolver</a></code>,
<code><a href='LassoSolver.html'>LassoSolver</a></code>, <code><a href='PearsonSolver.html'>PearsonSolver</a></code>,
<code><a href='RandomForestSolver.html'>RandomForestSolver</a></code>, <code><a href='RidgeSolver.html'>RidgeSolver</a></code>,
<code><a href='SpearmanSolver.html'>SpearmanSolver</a></code>, <code><a href='SqrtLassoSolver.html'>SqrtLassoSolver</a></code>.</p>
<p>CandidateFilter Class Objects</p>
<p>The <code><a href='CandidateFilter-class.html'>CandidateFilter</a></code> class is a base class that is generally used to filter
the transcription factors in the expression matrix to obtain a set of candidate
regulators. Filtering method depends on the filter type chosen; there are currently the
following subclasses: <code><a href='FootprintFilter-class.html'>FootprintFilter</a></code>, <code><a href='HumanDHSFilter-class.html'>HumanDHSFilter</a></code>,
<code><a href='GeneOntologyFilter-class.html'>GeneOntologyFilter</a></code>, and <code><a href='VarianceFilter-class.html'>VarianceFilter</a></code>. The filters are
applied using the <code><a href='getCandidates.html'>getCandidates</a></code> method on a given
<code><a href='CandidateFilter-class.html'>CandidateFilter</a></code> object.</p>
<p>FootprintFinder Class Objects</p>
<p>The  <code><a href='FootprintFinder-class.html'>FootprintFinder</a></code> class is designed to allow extraction
of gene footprinting information from existing PostgreSQL or SQLite
databases. In standard use of the `trena` package, it is used solely by
the <code><a href='getCandidates.html'>getCandidates</a></code> method for a <code><a href='FootprintFilter-class.html'>FootprintFilter</a></code>
object. However, a  <code><a href='FootprintFinder-class.html'>FootprintFinder</a></code> object has many more
available methods that allow it to extract information more flexibly.</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#details">Details</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Paul Shannon.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


